astrolabe/spec_runner.py

```diff
@@ -0,0 +1,400 @@
+        Initialize a cluster with the configuration required by the test
+        specification.
+        """
+        LOGGER.info("Initializing cluster {!r}".format(self.cluster_name))
+
+        cluster_config = self.spec.maintenancePlan.initial.\
+            clusterConfiguration.copy()
+        cluster_config["name"] = self.cluster_name
+        try:
+            self.client.groups[self.get_group().id].clusters.post(
+                **cluster_config)
+        except AtlasApiError as exc:
+            if exc.error_code == 'DUPLICATE_CLUSTER_NAME':
+                # Cluster already exists. Simply re-configure it.
+                # Cannot send cluster name when updating existing cluster.
+                cluster_config.pop("name")
+                self.client.groups[self.get_group().id].\
+                    clusters[self.cluster_name].patch(**cluster_config)
+
+        # Apply processArgs if provided.
+        process_args = self.spec.maintenancePlan.initial.processArgs
+        if process_args:
+            self.client.groups[self.get_group().id].\
+                clusters[self.cluster_name].processArgs.patch(**process_args)
+
+    def run(self, persist_cluster=False):
+        LOGGER.info("Running test {!r} on cluster {!r}".format(
+            self.id, self.cluster_name))
+
+        # Step-0: sanity-check the cluster configuration.
+        self.verify_cluster_configuration_matches("initial")
+
+        # Start the test timer.
+        timer = Timer()
+        timer.start()
+
+        # Step-1: load test data.
+        test_data = self.spec.driverWorkload.get('testData')
+        if test_data:
+            LOGGER.info("Loading test data on cluster {!r}".format(
+                self.cluster_name))
+            connection_string = self.get_connection_string()
+            client = MongoClient(connection_string, w="majority")
+            coll = client.get_database(
+                self.spec.driverWorkload.database).get_collection(
+                self.spec.driverWorkload.collection)
+            coll.drop()
+            coll.insert_many(test_data)
+            LOGGER.info("Successfully loaded test data on cluster {!r}".format(
+                self.cluster_name))
+
+        # Step-2: run driver workload.
+        LOGGER.info("Starting workload executor")
+        connection_string = self.get_connection_string()
+        driver_workload = json.dumps(self.spec.driverWorkload)
+        worker_subprocess = subprocess.Popen([
+            self.config.workload_executor, connection_string,
+            driver_workload], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        LOGGER.info("Started workload executor [PID: {}]".format(
+            worker_subprocess.pid))
+
+        # Step-3: begin maintenance routine.
+        final_config = self.spec.maintenancePlan.final
+        cluster_config = final_config.clusterConfiguration
+        process_args = final_config.processArgs
+
+        if not cluster_config and not process_args:
```

Comment:
should we consider doing validation much earlier, so you can bail before any I/O if need be?

Reply:
Good point. In general, we just need a lot more validation of the test specification. I have opened https://github.com/mongodb-labs/drivers-atlas-testing/issues/10 to do this.